streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStore.java

```diff
@@ -306,26 +319,41 @@ public synchronized boolean hasNext() {
                 } else {
                     next = KeyValue.pair(new Bytes(nextWithTimestamp), iterWithTimestamp.value());
                     nextWithTimestamp = null;
-                    iterWithTimestamp.next();
+                    if (reverse) {
+                        iterWithTimestamp.prev();
+                    } else {
+                        iterWithTimestamp.next();
+                    }
                 }
             } else {
                 if (nextWithTimestamp == null) {
                     next = KeyValue.pair(new Bytes(nextNoTimestamp), convertToTimestampedFormat(iterNoTimestamp.value()));
                     nextNoTimestamp = null;
-                    iterNoTimestamp.next();
+                    if (reverse) {
+                        iterNoTimestamp.prev();
+                    } else {
+                        iterNoTimestamp.next();
+                    }
                 } else {
                     if (comparator.compare(nextNoTimestamp, nextWithTimestamp) <= 0) {
```

Comment:
If `compare(noTimestamp, withTimestamp)` <= 0 then `withTimestamp >= noTimestamp`, so for the reverse case we would actually want to return `withTimestamp` next here (and vice versa in the else block)